ui/lib/core/addon/components/download-button.js

```diff
@@ -1,51 +1,67 @@
+/**
+ * @module DownloadButton
+ * DownloadButton components are an action button used to download data. Both the action text and icon are yielded.
+ *
+ * @example
+ * ```js
+ *   <DownloadButton
+ *     @buttonClass="toolbar-link"
+ *     @data={{this.data}}
+ *     @filename={{this.filename}}
+ *     @mime={{this.mime}}
+ *     @extension={{this.extension}}
+ *     @stringify={{true}}
+ *   >
+ *    <Icon @name="download" />
+ *      Download
+ *   </DownloadButton>
+ * ```
+ * @param {string} data - data to download
+ * @param {boolean} [stringify=false] - argument to stringify the data before passing to the File constructor
+ * @param {string} [filename] - name of file that prefixes the ISO timestamp generated when download
+ * @param {string} [mime='text/plain'] - media type to be downloaded
+ * @param {string} [extension='txt'] - file extension
+ * @param {string} [buttonClass] - override css class defaults to customize how the button looks
+ */
 
-export default Component.extend({
-  layout: hbs`{{#if (has-block)}} {{yield}} {{else}} {{actionText}} {{/if}}`,
-  tagName: 'a',
-  role: 'button',
-  attributeBindings: ['role', 'download', 'href'],
-  download: computed('filename', 'extension', function () {
-    return `${this.filename}-${new Date().toISOString()}.${this.extension}`;
-  }),
+export default class DownloadButton extends Component {
+  get extension() {
+    return this.args.extension || 'txt';
+  }
 
-  fileLike: computed('data', 'mime', 'stringify', 'download', function () {
-    let file;
-    let data = this.data;
-    const filename = this.download;
-    const mime = this.mime;
-    if (this.stringify) {
-      data = JSON.stringify(data, null, 2);
-    }
-    if (window.navigator.msSaveOrOpenBlob) {
-      file = new Blob([data], { type: mime });
-      file.name = filename;
-    } else {
-      file = new File([data], filename, { type: mime });
-    }
-    return file;
-  }),
+  get mime() {
+    return this.args.mime || 'text/plain';
+  }
 
-  href: computed('fileLike', function () {
-    return window.URL.createObjectURL(this.fileLike);
-  }),
+  get filename() {
+    const defaultFilename = `${new Date().toISOString()}.${this.extension}`;
+    return this.args.filename ? this.args.filename + '-' + defaultFilename : defaultFilename;
+  }
 
-  click(event) {
-    if (!window.navigator.msSaveOrOpenBlob) {
-      return;
+  get data() {
+    if (this.args.stringify) {
+      return JSON.stringify(this.args.data, null, 2);
     }
-    event.preventDefault();
-    const file = this.fileLike;
-    //lol whyyyy
-    window.navigator.msSaveOrOpenBlob(file, file.name);
-  },
+    return this.args.data;
+  }
+
+  get content() {
+    const content = new File([this.data], this.filename, { type: this.mime });
+    return content;
+  }
 
-  actionText: 'Download',
-  data: null,
-  filename: null,
-  mime: 'text/plain',
-  extension: 'txt',
-  stringify: false,
-});
+  // TODO refactor and call service instead
```

Comment:
Would this be the `DownloadCsvService`?

Reply:
Yep! Chelsea is refactoring the service now to be more general. I'll then update this action to use the service when I use this component to implement downloading a generated PKI key